src/wiringx.c

```diff
@@ -400,90 +519,111 @@ EXPORT int wiringXI2CSetup(const char *path, int devId) {
 	return fd;
 }
 
-EXPORT int wiringXSPIGetFd(int channel) {
-	return spi[channel & 1].fd;
+EXPORT int wiringXSPIGetFd(int handle) {
+	struct spi_t *state = NULL;
+
+	if(handle < 0 || handle >= nrspilist) {
+		wiringXLog(LOG_ERR, "wiringX trying to use an unknown SPI connection handle: %i", handle);
+		return -1;
+	}
+
+	state = &spilist[handle].data;
+	return state->fd;
 }
 
-EXPORT int wiringXSPIDataRW(int channel, unsigned char *data, int len) {
+EXPORT int wiringXSPIDataRW(int handle, unsigned char *data, int len) {
+	struct spi_t *state = NULL;
 	struct spi_ioc_transfer tmp;
+
+	if(handle < 0 || handle >= nrspilist) {
+		wiringXLog(LOG_ERR, "wiringX trying to use an unknown SPI connection handle: %i", handle);
+		return -1;
+	}
+
+	state = &spilist[handle].data;
 	memset(&tmp, 0, sizeof(tmp));
-	channel &= 1;
 
 	tmp.tx_buf = (uintptr_t)data;
 	tmp.rx_buf = (uintptr_t)data;
 	tmp.len = len;
-	tmp.delay_usecs = spi[channel].delay;
-	tmp.speed_hz = spi[channel].speed;
-	tmp.bits_per_word = spi[channel].bits_per_word;
+	tmp.delay_usecs = state->delay;
+	tmp.speed_hz = state->speed;
+	tmp.bits_per_word = state->bits_per_word;
 #ifdef SPI_IOC_WR_MODE32
 	tmp.tx_nbits = 0;
 #endif
 #ifdef SPI_IOC_RD_MODE32
 	tmp.rx_nbits = 0;
 #endif
 
-	if(ioctl(spi[channel].fd, SPI_IOC_MESSAGE(1), &tmp) < 0) {
-		wiringXLog(LOG_ERR, "wiringX is unable to read/write from channel %d (%s)", channel, strerror(errno));
+	if(ioctl(state->fd, SPI_IOC_MESSAGE(1), &tmp) < 0) {
+		wiringXLog(LOG_ERR, "wiringX is unable to read/write from SPI device /dev/spidev%u.%u (%s)", spilist[handle].id.device, spilist[handle].id.channel, strerror(errno));
 		return -1;
 	}
 	return 0;
 }
 
-EXPORT int wiringXSPISetup(int channel, int speed) {
-	const char *device = NULL;
-
-	channel &= 1;
+EXPORT int wiringXSPISetup(uint8_t device, uint8_t channel, int speed) {
+	// maximum length of path to spidev: "/dev/spidev255.255\0"
+	char spidev[19] = { 0 };
+	struct spi_id_t id = { device, channel };
+	struct spi_t state;
```

Comment:
Can you memset the `state`

Reply:
sure, = {0};

Reply:
Yes, I am sure.
See this [stackoverflow answer](https://stackoverflow.com/a/11152199).
By using {0} we initialize all members of the struct to 0 unconditionally. And we avoid a call to memset and let the compiler take care of this.